In [1]:
import os
import numpy as np
import pandas as pd
pd.set_option('display.max_colwidth', None)

### First table is S0101 - Age and Sex

In [2]:
# path to the S0101 subject table
data_dir = 'data/subject_tables/unzipped_files/ACSST5Y2019.S0101_2021-01-12T120250'

In [3]:
S0101_data = pd.read_csv(os.path.join(data_dir, 'ACSST5Y2019.S0101_data_with_overlays_2021-01-08T174020.csv'), low_memory=False)
S0101_metadata = pd.read_csv(os.path.join(data_dir, 'ACSST5Y2019.S0101_metadata_2021-01-08T174020.csv'))

In [4]:
print(f'Shape of data table proper is {S0101_data.shape} \n'
    + f'Shape of metadata table is {S0101_metadata.shape}')

Shape of data table proper is (5266, 458) 
Shape of metadata table is (457, 2)


In [5]:
S0101_data.head(2)

GEO_ID                                       NAME  \
0                    id                       Geographic Area Name   
1  1400000US48001950100  Census Tract 9501, Anderson County, Texas   

                      S0101_C01_001E  \
0  Estimate!!Total!!Total population   
1                               4844   

                             S0101_C01_001M  \
0  Margin of Error!!Total!!Total population   
1                                       524   

                                          S0101_C01_002E  \
0  Estimate!!Total!!Total population!!AGE!!Under 5 years   
1                                                    349   

                                                 S0101_C01_002M  \
0  Margin of Error!!Total!!Total population!!AGE!!Under 5 years   
1                                                           131   

                                         S0101_C01_003E  \
0  Estimate!!Total!!Total population!!AGE!!5 to 9 years   
1                                                   269   

                                                S0101_C01_003M  \
0  Margin of Error!!Total!!Total population!!AGE!!5 to 9 years   
1                                                          119   

                                           S0101_C01_004E  \
0  Estimate!!Total!!Total population!!AGE!!10 to 14 years   
1                                                     372   

                                                  S0101_C01_004M  ...  \
0  Margin of Error!!Total!!Total population!!AGE!!10 to 14 years  ...   
1                                                            137  ...   

                                                                         S0101_C06_034E  \
0  Estimate!!Percent Female!!Total population!!SUMMARY INDICATORS!!Age dependency ratio   
1                                                                                   (X)   

                                                                                S0101_C06_034M  \
0  Margin of Error!!Percent Female!!Total population!!SUMMARY INDICATORS!!Age dependency ratio   
1                                                                                          (X)   

                                                                             S0101_C06_035E  \
0  Estimate!!Percent Female!!Total population!!SUMMARY INDICATORS!!Old-age dependency ratio   
1                                                                                       (X)   

                                                                                    S0101_C06_035M  \
0  Margin of Error!!Percent Female!!Total population!!SUMMARY INDICATORS!!Old-age dependency ratio   
1                                                                                              (X)   

                                                                           S0101_C06_036E  \
0  Estimate!!Percent Female!!Total population!!SUMMARY INDICATORS!!Child dependency ratio   
1                                                                                     (X)   

                                                                                  S0101_C06_036M  \
0  Margin of Error!!Percent Female!!Total population!!SUMMARY INDICATORS!!Child dependency ratio   
1                                                                                            (X)   

                                                       S0101_C06_037E  \
0  Estimate!!Percent Female!!Total population!!PERCENT ALLOCATED!!Sex   
1                                                                 (X)   

                                                              S0101_C06_037M  \
0  Margin of Error!!Percent Female!!Total population!!PERCENT ALLOCATED!!Sex   
1                                                                        (X)   

                                                       S0101_C06_038E  \
0  Estimate!!Percent Female!!Total population!!PERCENT ALLOCATED!!Age   
1                            

In [6]:
S0101_metadata.head(2)

GEO_ID                                 id
0            NAME               Geographic Area Name
1  S0101_C01_001E  Estimate!!Total!!Total population

In [7]:
# A dictionary to keep track of the original columns and the first row of df which is text explanation of column values.
original_column_dictionary = dict(zip(S0101_data.columns, S0101_data.iloc[0,:]))

We should break apart the GEOID field into its FIPS components for easier matching of datasets.

FIPS codes are 11 digits long. 2 state digits, 3 county digits, 6 tract digits.

In the GEO_ID column the FIPS code comes after the `US` characters.

In [8]:
# Breaking apart GEO_ID column into the FIPS components and creating new columns for them.
# Split breaks the string into a list and then we specify the second part and which chars we want from that part
# the conditional logic is because of the pesky 'id' value in row 0, we keep it for now for easier joins later.

S0101_data["STATE"] = S0101_data.GEO_ID.apply(lambda x: x.split('US')[1][:2] if len(x) > 3 else 'state')
S0101_data["COUNTYFP"] = S0101_data.GEO_ID.apply(lambda x: x.split('US')[1][2:5] if len(x) > 3 else 'countyfp')
S0101_data["TRACTCE"] = S0101_data.GEO_ID.apply(lambda x: x.split('US')[1][5:] if len(x) > 3 else 'tractce')

Note: these will have iloc indicies of 458, 459, and 460. Important to remember for later.

In [9]:
# Visual check to make sure these are properly formatted
S0101_data[['STATE', 'COUNTYFP', 'TRACTCE']].head(5)

STATE COUNTYFP TRACTCE
0   NaN      NaN     NaN
1    48      001  950100
2    48      001  950401
3    48      001  950402
4    48      001  950500

The rows in the metadata table provide a mapping to the columns in the actual data table, but there are some minor offsets
to be made in order to make matching indices (i.e. row 1 in metadata table corresponds to column 1).

If we drop the GEO_ID and NAME columns in the dataset and drop the first row of the metadataset then they will be aligned.

In [10]:
S0101_data.drop(columns=['GEO_ID', 'NAME'], inplace=True)
S0101_metadata.drop(index=0, axis=0, inplace=True)
S0101_metadata.reset_index(drop=True, inplace=True)

# Subtracting 3 to S0101_data.shape[1] to account for the our creation of 3 new columns
assert (S0101_data.shape[1]-3 == S0101_metadata.shape[0]), print('mismatch in metadata and data correspondence')

In [11]:
int_locations = [3, *(list(range(15,30,2))), 43, 47, 61, 155, 307] # These are the rows as they appear in metadata excel sheet
int_locations = [x-3 for x in int_locations] # Offsetting these indicies to account for drops, header and 0-index pandas scheme

In [12]:
# Checking to make sure we have specified the columns we want.
S0101_metadata.iloc[int_locations]

GEO_ID  \
0    S0101_C01_001E   
12   S0101_C01_007E   
14   S0101_C01_008E   
16   S0101_C01_009E   
18   S0101_C01_010E   
20   S0101_C01_011E   
22   S0101_C01_012E   
24   S0101_C01_013E   
26   S0101_C01_014E   
40   S0101_C01_021E   
44   S0101_C01_023E   
58   S0101_C01_030E   
152  S0101_C03_001E   
304  S0101_C05_001E   

                                                                                id  
0                                                Estimate!!Total!!Total population  
12                          Estimate!!Total!!Total population!!AGE!!25 to 29 years  
14                          Estimate!!Total!!Total population!!AGE!!30 to 34 years  
16                          Estimate!!Total!!Total population!!AGE!!35 to 39 years  
18                          Estimate!!Total!!Total population!!AGE!!40 to 44 years  
20                          Estimate!!Total!!Total population!!AGE!!45 to 49 years  
22                          Estimate!!Total!!Total population!!AGE!!50 to 54 years  
24                          Estimate!!Total!!Total population!!AGE!!55 to 59 years  
26                          Estimate!!Total!!Total population!!AGE!!60 to 64 years  
40      Estimate!!Total!!Total population!!SELECTED AGE CATEGORIES!!15 to 17 years  
44      Estimate!!Total!!Total population!!SELECTED AGE CATEGORIES!!18 to 24 years  
58   Estimate!!Total!!Total population!!SELECTED AGE CATEGORIES!!65 years and over  
152                                               Estimate!!Male!!Total population  
304                                             Estimate!!Female!!Total population

In [13]:
# Creating a list of the column names we want so we can use this to filter our data set.
wanted_columns = S0101_metadata.GEO_ID.iloc[int_locations].to_list()

In [14]:
# Need to account for the 3 columns we created by breaking up GEO_ID that have no mapping to metdata df.
additional_columns = list(S0101_data.columns[-3:])
wanted_columns.extend(additional_columns)

In [15]:
wanted_columns

['S0101_C01_001E',
 'S0101_C01_007E',
 'S0101_C01_008E',
 'S0101_C01_009E',
 'S0101_C01_010E',
 'S0101_C01_011E',
 'S0101_C01_012E',
 'S0101_C01_013E',
 'S0101_C01_014E',
 'S0101_C01_021E',
 'S0101_C01_023E',
 'S0101_C01_030E',
 'S0101_C03_001E',
 'S0101_C05_001E',
 'STATE',
 'COUNTYFP',
 'TRACTCE']

In [16]:
S0101_data = S0101_data[wanted_columns]

In [17]:
S0101_data.head(5)

S0101_C01_001E  \
0  Estimate!!Total!!Total population   
1                               4844   
2                               4838   
3                               7511   
4                               4465   

                                           S0101_C01_007E  \
0  Estimate!!Total!!Total population!!AGE!!25 to 29 years   
1                                                     252   
2                                                     843   
3                                                     586   
4                                                     483   

                                           S0101_C01_008E  \
0  Estimate!!Total!!Total population!!AGE!!30 to 34 years   
1                                                     197   
2                                                     907   
3                                                    1339   
4                                                     199   

                                           S0101_C01_009E  \
0  Estimate!!Total!!Total population!!AGE!!35 to 39 years   
1                                                     335   
2                                                     753   
3                                                    1333   
4                                                     268   

                                           S0101_C01_010E  \
0  Estimate!!Total!!Total population!!AGE!!40 to 44 years   
1                                                     163   
2                                                     444   
3                                                    1171   
4                                                     413   

                                           S0101_C01_011E  \
0  Estimate!!Total!!Total population!!AGE!!45 to 49 years   
1                                                     312   
2                                                     419   
3                                                     980   
4                                                     204   

                                           S0101_C01_012E  \
0  Estimate!!Total!!Total population!!AGE!!50 to 54 years   
1                                                     447   
2                                                     314   
3                                                     790   
4                                                     224   

                                           S0101_C01_013E  \
0  Estimate!!Total!!Total population!!AGE!!55 to 59 years   
1                                                     318   
2                                                     223   
3                                                     485   
4                                                     253   

                                           S0101_C01_014E  \
0  Estimate!!Total!!Total population!!AGE!!60 to 64 years   
1                                                     318   
2                                                      73   
3                                                     315   
4                                                     191   

                                                               S0101_C01_021E  \
0  Estimate!!Total!!Total population!!SELECTED AGE CATEGORIES!!15 to 17 years   
1                                                                         250   
2                                                                           0   
3                                                                           5   
4                                                                         190   

                                                               S0101_C01_023E  \
0  Estimate!!Total!!Total population!!SELECTED AGE CATEGORIES!!18 to 24 years   
1                                                                         205   
2                                                                         634   
3                                 

In [18]:
new_column_dictionary = dict(zip(S0101_data.columns, S0101_data.iloc[0]))

In [19]:
S0101_data.drop(index=0, axis=0, inplace=True)
S0101_data.reset_index(drop=True, inplace=True)

In [22]:
S0101_data.shape

(5265, 17)

Reorder the columns and then compress the bins if necessary to align with our wanted columns.

In [23]:
curr_columns = list(S0101_data.columns)

In [30]:
for i in curr_columns:
    print(f'Column {i} is {new_column_dictionary[i]}')

Column S0101_C01_001E is Estimate!!Total!!Total population
Column S0101_C01_007E is Estimate!!Total!!Total population!!AGE!!25 to 29 years
Column S0101_C01_008E is Estimate!!Total!!Total population!!AGE!!30 to 34 years
Column S0101_C01_009E is Estimate!!Total!!Total population!!AGE!!35 to 39 years
Column S0101_C01_010E is Estimate!!Total!!Total population!!AGE!!40 to 44 years
Column S0101_C01_011E is Estimate!!Total!!Total population!!AGE!!45 to 49 years
Column S0101_C01_012E is Estimate!!Total!!Total population!!AGE!!50 to 54 years
Column S0101_C01_013E is Estimate!!Total!!Total population!!AGE!!55 to 59 years
Column S0101_C01_014E is Estimate!!Total!!Total population!!AGE!!60 to 64 years
Column S0101_C01_021E is Estimate!!Total!!Total population!!SELECTED AGE CATEGORIES!!15 to 17 years
Column S0101_C01_023E is Estimate!!Total!!Total population!!SELECTED AGE CATEGORIES!!18 to 24 years
Column S0101_C01_030E is Estimate!!Total!!Total population!!SELECTED AGE CATEGORIES!!65 years and ove

In [31]:
# Manual reordering might be necessary here
new_columns = ['S0101_C01_001E',
    'S0101_C01_021E',
    'S0101_C01_023E',
    'S0101_C01_007E',
    'S0101_C01_008E',
    'S0101_C01_009E',
    'S0101_C01_010E',
    'S0101_C01_011E',
    'S0101_C01_012E',
    'S0101_C01_013E',
    'S0101_C01_014E', 
    'S0101_C01_030E',
    'S0101_C03_001E',
    'S0101_C05_001E',
    'STATE',
    'COUNTYFP',
    'TRACTCE']

In [32]:
for i in new_columns:
    print(f'Column {i} is {new_column_dictionary[i]}')

Column S0101_C01_001E is Estimate!!Total!!Total population
Column S0101_C01_021E is Estimate!!Total!!Total population!!SELECTED AGE CATEGORIES!!15 to 17 years
Column S0101_C01_023E is Estimate!!Total!!Total population!!SELECTED AGE CATEGORIES!!18 to 24 years
Column S0101_C01_007E is Estimate!!Total!!Total population!!AGE!!25 to 29 years
Column S0101_C01_008E is Estimate!!Total!!Total population!!AGE!!30 to 34 years
Column S0101_C01_009E is Estimate!!Total!!Total population!!AGE!!35 to 39 years
Column S0101_C01_010E is Estimate!!Total!!Total population!!AGE!!40 to 44 years
Column S0101_C01_011E is Estimate!!Total!!Total population!!AGE!!45 to 49 years
Column S0101_C01_012E is Estimate!!Total!!Total population!!AGE!!50 to 54 years
Column S0101_C01_013E is Estimate!!Total!!Total population!!AGE!!55 to 59 years
Column S0101_C01_014E is Estimate!!Total!!Total population!!AGE!!60 to 64 years
Column S0101_C01_030E is Estimate!!Total!!Total population!!SELECTED AGE CATEGORIES!!65 years and ove

In [33]:
# Reorder columns using updated order
S0101_data = S0101_data[new_columns]